Feature/cldsrv 515 quotas

williamlardier opened the pull request:
Adding quota evaluation logic in S3:
- Both accounts (from authz results) and buckets
- Use a Scubaclient instance to get the latest metrics with inflight metrics (i.e., ongoing bytes not yet acknowledged) 
- Periodic healthchecks to dynamically handle problems with the quota service (today, we disable quota checks)
- Add configuration for `scubaclient` (host & port)
- Support WRITE and DELETE APIs with inflights. Copy APIs are handled slightly differently as we need to get the original object size.
- Unit tests
- Functional tests (no account coverage, to be done during integration testing)

> Note: won't pass the CI because we cannot install scubaclient without setup in the dockerfile, or having scubaclient public.

bert-e commented:
# Hello williamlardier,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://bert-e.scality.net/scality/cloudserver/doc/user).





<details>
  <summary><b>Available options</b></summary>

  name   | description  | privileged   | authored
  ------ | ------------ | ------------ |----------
  `/after_pull_request` | Wait for the given pull request id to be merged before continuing with the current one. |  | 
  `/bypass_author_approval` | Bypass the pull request author's approval |  :star:  | 
  `/bypass_build_status` | Bypass the build and test status |  :star:  | 
  `/bypass_commit_size` | Bypass the check on the size of the changeset ```TBA``` |  :star:  | 
  `/bypass_incompatible_branch` | Bypass the check on the source branch prefix |  :star:  | 
  `/bypass_jira_check` | Bypass the Jira issue check |  :star:  | 
  `/bypass_peer_approval` | Bypass the pull request peers' approval |  :star:  | 
  `/bypass_leader_approval` | Bypass the pull request leaders' approval |  :star:  | 
  `/approve` | Instruct Bert-E that the author has approved the pull request. |  |  :writing_hand: 
  `/create_pull_requests` | Allow the creation of integration pull requests. |  | 
  `/create_integration_branches` | Allow the creation of integration branches. |  | 
  `/no_octopus` | Prevent Wall-E from doing any octopus merge and use multiple consecutive merge instead |  | 
  `/unanimity` | Change review acceptance criteria from `one reviewer at least` to `all reviewers` |  | 
  `/wait` | Instruct Bert-E not to run until further notice. |  | 
  

</details>



<details>
  <summary><b>Available commands</b></summary>

  name   | description  | privileged
  ------ | ------------ | ------------
  `/help` | Print Bert-E's manual in the pull request. | 
  `/status` | Print Bert-E's current status in the pull request ```TBA``` | 
  `/clear` | Remove all comments from Bert-E from the history ```TBA``` | 
  `/retry` | Re-start a fresh build ```TBA``` | 
  `/build` | Re-start a fresh build ```TBA``` | 
  `/force_reset` | Delete integration branches & pull requests, and restart merge process from the beginning. | 
  `/reset` | Try to remove integration branches unless there are commits on them which do not appear on the source branch. | 
  
</details>




*Status report is not available.*

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue CLDSRV-515 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.8.21`


Please check the `Fix Version/s` of CLDSRV-515, or the target
branch of this pull request.

ghost commented:
I have a question for you @tmacro : is scubaclient meant to become public at some point? If we use it here, it might break the service for the open source community. It also breaks the dockerfile builds, unless we pass the github token in it...

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

williamlardier commented on tests/utilities/mock/Scuba.js:
```diff
@@ -0,0 +1,95 @@
+            }, timeout);
+        } else {
+            if (this._data.bucket[bucket]) {
+                this._data.bucket[bucket].inflight += inflight;
```
This is fixed.

williamlardier commented on tests/utilities/mock/Scuba.js:
```diff
@@ -0,0 +1,95 @@
+        if (timeout) {
+            setTimeout(() => {
+                if (this._data.bucket[bucket]) {
+                    this._data.bucket[bucket].inflight += inflight;
```
This is fixed.

francoisferrand commented on .github/workflows/tests.yaml:
```diff
@@ -363,6 +363,40 @@ jobs:
+      JOB_NAME: ${{ github.job }}
+    steps:
+    - name: Checkout
+      uses: actions/checkout@v3
```
```suggestion
      uses: actions/checkout@v4
```

williamlardier replied:
https://github.com/scality/cloudserver/pull/5554/commits/0d4cc38000c41e2f53946a1b9ec225eab16d1e2a

Seragonia committed:
CLDSRV-515: bump arsenal and vaultclient, introduce scubaclient

francoisferrand commented on .github/workflows/tests.yaml:
```diff
@@ -363,6 +363,40 @@ jobs:
+        bash wait_for_local_port.bash 8000 40
+        yarn run test_scuba | tee /tmp/artifacts/${{ github.job }}/tests.log
+    - name: Upload logs to artifacts
+      uses: scality/action-artifacts@v3
```
@v4

francoisferrand commented on .github/workflows/tests.yaml:
```diff
@@ -363,6 +363,40 @@ jobs:
+      SCUBA_HOST: localhost
+      SCUBA_PORT: 8100
+      SCUBA_HEALTHCHECK_FREQUENCY: 100
+      CLOUDSERVER_IMAGE: ghcr.io/${{ github.repository }}/cloudserver:${{ github.sha }}
```
ghcr.io/${{ github.repository }}:${{ github.sha }}

williamlardier replied:
I updated the actions.

williamlardier replied:
Actually, no, I cannot make this change currently, the PR is based on another one, that is not yet up to date with ghcr migration, so the image in the CI still is the one you commented, with /cloudserver

williamlardier replied:
Done after rebasing.

francoisferrand commented on Dockerfile:
```diff
@@ -23,6 +23,7 @@ RUN apt-get update \
 
 ENV PYTHON=python3
 COPY package.json yarn.lock /usr/src/app/
+RUN npm install typescript -g
```
why do we need to install typescript ?
should it not be already installed by `yarn install` if needed ?

williamlardier replied:
When building the sub dependencies, especially scubaclient, we have an error where tsc is not found (for scubaclient), because we need to install tsc globally first. I did not try to dig into it much, but it's pretty common to install tsc globally in dockerfiles of ts projects (or projects using it).

francoisferrand commented on lib/api/apiUtils/authorization/permissionChecks.js:
```diff
@@ -372,6 +381,141 @@ function processBucketPolicy(requestType, bucket, canonicalID, arn, bucketOwner,
+
+    return async.forEach(apiNames, (apiName, done) => {
+        let shouldEvaluateCopyObject = false;
+        if (apiName === 'objectGet' && (apiMethod === 'objectCopy' || apiMethod === 'objectPutCopyPart')) {
```
should we not have maps similar to `actionWithDataDeletion` for detecting this case?

williamlardier replied:
yes, but not in arsenal, in the cloudserver's consts I think, this is not cross service.

francoisferrand replied:
theorically maybe, but practically it could make sense in arsenal as well:
* arsenal already includes lots of cloudserver (or backbeat, or vault) specific code ;
* it may make sense to keep this together with the other 2 variables, since this whole algorithm is based on these variables. The question is really "_would it make sense to modify just this new variable without considering if an update is required on the other 2_" ?

williamlardier replied:
After reviewing this, I'm not sure anymore it makes sense as a map: we need to pass two strings, the current API and the current action we evaluate, so better having a small function; a map won't make a lot of sense... The function will be added as a new function in arsenal.

francoisferrand commented on lib/api/apiUtils/authorization/permissionChecks.js:
```diff
@@ -372,6 +381,141 @@ function processBucketPolicy(requestType, bucket, canonicalID, arn, bucketOwner,
+                        action: apiName,
+                        inflight,
+                    }, (err, bucketMetrics) => {
+                        if (err) {
+                            return parallelDone(err);
+                        }
+                        if (bucketMetrics.bytesTotal > bucketQuota) {
```
nit:
```suggestion
                        if (!err && bucketMetrics.bytesTotal > bucketQuota) {
```

francoisferrand commented on lib/api/apiUtils/authorization/permissionChecks.js:
```diff
@@ -372,6 +381,141 @@ function processBucketPolicy(requestType, bucket, canonicalID, arn, bucketOwner,
+                return done(err);
+            }
+            if (!actionWithDataDeletion[apiMethod] && (bucketQuotaExceeded || accountQuotaExceeded)) {
+                if (apiMethod?.endsWith('Delete')) {
```
seems hard-coded, can we do better: something like

```suggestion
                if (actionWithDataDeletion[apiMethod]) {
```

williamlardier replied:
Yes, missed this one, thx

francoisferrand commented on tests/scuba/awsNodeSdk.js:
should this not be a _quota_ folder instead of _scuba_ ?

williamlardier replied:
Yes, if we have more use cases later, we can rename it. And maybe it'll be removed before, if we run the real scuba.

francoisferrand commented on tests/unit/auth/permissionChecks.js:
```diff
@@ -534,3 +554,307 @@ describe('validatePolicyConditions', () => {
+        sinon.restore();
+    });
+
+    it('should return null if quota is <= 0 or scuba is disabled', done => {
```
```suggestion
    it('should return null if quota is <= 0', done => {
```

francoisferrand commented on tests/unit/auth/permissionChecks.js:
```diff
@@ -534,3 +554,307 @@ describe('validatePolicyConditions', () => {
+        });
+    });
+
+    it('should return null if metrics retrieval fails', done => {
```
side note (not for this PR in any case) : bypassing quota in that case may allow for easily bypassing the quotas with DOS attack... e.g. if we try to hammer the ser-ver with many putObject requests, scuba will probably fail before cloudserver.... and the request may bypass quota :/

williamlardier replied:
It can be worth discussing this PR. We have two things:
- The periodic healthchecks, they are the one that determine the state of the quota checks.
- The error handling, when evaluating quotas. When there is no quota and scuba is enabled, it means count items never ran for the current bucket/account, and thus we need to ignore the quotas. but for the other cases.

Today we ignore the quotas if the healthchecks failed, or if there is an error when getting the metrics (without discriminating between really missing metrics and other issues).

If the bucket/quota exists, we could try to still return an error in this case, but I'm not sure how.

But anyway, the scuba service must be sized properly, so that it is not a bottleneck. Mongo can be, but scuba shouldn't...

francoisferrand commented on .github/workflows/tests.yaml:
```diff
@@ -363,6 +363,40 @@ jobs:
+      run: |-
+        set -ex -o pipefail;
+        bash wait_for_local_port.bash 8000 40
+        yarn run test_scuba | tee /tmp/artifacts/${{ github.job }}/tests.log
```
even with the function tests, we actually don't test with the "real" scuba : so we may easily introduce issue if the mock does not match the actual scuba API... → we should find a way to run these tests with either/both mock scuba or "real" scuba
(i.e. get feedback earlier than when it reaches zenko...)

williamlardier replied:
The functional tests are only testing cloudserver contract here, but yes, to me this is a second step (as we'll have the integration test in Zenko), in the same way we test Vault + MongoDB today, we can really test with scuba, what do you think?

williamlardier replied:
Created https://scality.atlassian.net/browse/CLDSRV-528, in the improvement epic. It will be about not only scuba, but also vault (we don't test it today in the CI...)

francoisferrand commented on lib/metadata/metadataUtils.js:
```diff
@@ -238,6 +261,15 @@ function standardMetadataValidateBucketAndObj(params, actionImplicitDenies, log,
+            }
+            contentLength = processBytesToWrite(request, bucket, versionId, contentLength, objMD);
+            return validateQuotas(bucket, request.accountQuotas, requestType, request.apiMethod,
+                contentLength, log, err => next(err, bucket, objMD));
```
Putting the quota check here will impact many APIs, introducing a very slight risk (like quota unexpectedly applied on read...), as well as negigible cost (extra test), but also possibly leading to unexpected error : if an invalid request is made (like restore on a hot object), it may thus get a quota error instead of the invalid state error...

Wondering what is the expected behavior, and/or if we should change it now or in a followup ticket.

williamlardier replied:
As discussed, we will, to better centralize the logic and minimize the risks, keep the quota evaluation logic as part of the authz checks. But:

- We will handle the restore object separately, as extending the restoratin duration means no quota change.
- We will revert the inflights set in case of error at the end of an API call, if needed.

williamlardier replied:
Changes are done:

- Generalized the quota logic implementation, it can be another implementation, not always "scuba" (which is one of the possible solutions from S3 point of view). The backend is to be specifiec with `S3QUOTA`. To be also more future-proof, by default, the inflights are disabled, as this will be our nominal use case later. A configuration is thus added under `quota` to support it (ZKOP is also updated).

- To handle errors in APIs, the quota evaluation logic modifies the request object and adds some new methods that are evaluated, for each action of the API, at the end, in case of error. For example, it ensures we do not change the inflights (when enabled) if we try to delete a locked object. -**Note**: I found that we always pass the `objectReplicate` action in the route backbeat, so quotas are in fact **not** evaluated in there. This will be updated later based on the decision we make.

- The quota logic is better filtered and will only execute when matching the list of known APIs, or if we pass specific flags (e.g., during copy operation).

francoisferrand commented on lib/Config.js:
```diff
@@ -1117,6 +1117,47 @@ class Config extends EventEmitter {
+        }
+        if (process.env.QUOTA_MAX_STALENESS_MS) {
+            assert(Number.isInteger(Number(process.env.QUOTA_MAX_STALENESS_MS)),
+                'bad config: quota.maxStalenessMS must be an integer');
```
```suggestion
                'bad config: QUOTA_MAX_STALENESS_MS must be an integer');
```

francoisferrand replied:
or better yet, merge this with the previous case, so we do this only once. something like:

```
const maxStaleness = process.env.QUOTA_MAX_STALENESS_MS || config.quota?.maxStatenessMS || 24 * 60 * 60 * 1000;
assert(Number.isInteger(maxStaleness), 'bad config: maxStalenessMS must be an integer');
this.quota = { maxStaleness };
```

williamlardier replied:
Updated

francoisferrand commented on lib/Config.js:
```diff
@@ -1117,6 +1117,47 @@ class Config extends EventEmitter {
+                assert(Number.isInteger(config.quota.maxStalenessMS),
+                    'bad config: quota.maxStalenessMS must be an integer');
+                this.quota = { maxStalenessMS: config.quota.maxStalenessMS };
+            }
```
in the `else` case, this.quota is not initialized...

francoisferrand commented on lib/api/apiUtils/authorization/permissionChecks.js:
```diff
@@ -372,6 +382,141 @@ function processBucketPolicy(requestType, bucket, canonicalID, arn, bucketOwner,
+                        if (err) {
+                            return parallelDone(err);
+                        }
+                        if (bucketMetrics.bytesTotal > bucketQuota) {
```
we should check here as well that the data returned by Scuba is not stale (as the quota request will return the date as well)

williamlardier replied:
The logic was updated with the `isMetricStale` method

Seragonia committed:
CLDSRV-515: do not recreate variable at every authz

Seragonia committed:
CLDSRV-515: add configuration for quotas

- Quota service is generic. We only support scuba backend now,
  but we can add others later, if needed, as long as they share
  the same implementation as the scuba client.
- Scuba configuration is passed for the scubaclient tool.
- Ability to disable the inflights is provided. This changes the
  behavior of the quota checks, so that the inflights won't be
  part of the request to the utilization metrics services. This
  reduces the complexity of the quota evaluation logic in case
  of error, as no cleanup will be needed in this case. This,
  however, requires a backend that can provide up to date metrics
  (i.e., <2s).

Seragonia committed:
CLDSRV-515: create a wrapper for scubaclient and quota service

Seragonia committed:
CLDSRV-515: bootstrap scuba on startup

Seragonia committed:
CLDSRV-515: implement the quota logic as an helper file

Seragonia committed:
CLDSRV-515: evaluate quotas

Quotas are evaluated:
- As part of the authorization process, after both the bucket and
  the object are authorized. The checks are skipped if the API does
  not need any quota evaluation, if the inflight bytes are 0 (i.e.,
  no data added, so no need to check the quota).
- The Copy APIs will evaluate the quotas when the source object is
  checked. In this particular case, the action is objectGet, so a
  flag is passed to force the quota evaluation logic. A subsequent
  check is done in the logic.
- The restoreObject API has a special case where the extension of
  the restoration duration would still cause the evaluation of the
  quotas, causing a potential increase in the inflights stored. We
  detect this case and remove any added inflight.

Seragonia committed:
CLDSRV-515: clear inflights in case of quota exceeded

- If the quotas are evaluated with success and inflights are
  enabled, it means the quota service will store the information
  and persist it till the next update of the utilization metrics.
  In this case, aany API that will fail after authorization would
  still mean that the bytes are considered, even if nothing was
  written. To overcome that, we call a function from the quota
  evaluation logic to erase anything that we wrote during the
  authorization.

francoisferrand commented on lib/api/apiUtils/object/objectRestore.js:
```diff
@@ -142,6 +143,12 @@ function objectRestore(metadata, mdUtils, userInfo, request, log, callback) {
                     'POST', bucketName, '200', 'restoreObject');
+                // If quotas are evaluated, increasing the quota duration should
+                // not affect the utilization metrics.
+                if (request.clearQuotaUpdatesForAPI?.[mdValueParams]) {
```
* the name of this attribute should be improved, not very clear at the moment
* I don't see where this variable is set... 🧐
* since this is set in each request, I guess we should be able to directly set the appropriate callback (which would be cleaner IMO)... but then I did not see how it is initalized or added to the request

williamlardier replied:
> since this is set in each request, I guess we should be able to directly set the appropriate callback

I am not sure we can, the variable is set during the quota evaluation, and each API has custom ways to call their final callbacks, depending on the HTTP verb, etc.

francoisferrand commented on .github/workflows/tests.yaml:
```diff
@@ -365,6 +365,42 @@ jobs:
+    env:
+      S3BACKEND: mem
+      S3QUOTA: scuba
+      QUOTA_ENABLE_INFLIGHTS: "true"
```
in-flights are always enabled in tests,  but disabled by default in Cloudserver: so we are introducing risk (aka debt, somehow...), which may hit us once we start using cloudserver with no inflight...

If we do not enable inflight processing by default, this case should be tested as well to avoid future problems. (or we could stick with the inflight enabled for now, even if ZKOP/... force enable the flag, to allow us to later disable it...)

williamlardier replied:
We have (unit) test covering when there is no inflight enabled. Not for the functional tests, because the current scuba + mock both use the inflights, I'm not sure this introduces any debt...

If we enable inflights by default, and we someday deploy a real scuba in the CI, we will still have debt when inflights are dropped. If we disable inflights by default, this is the same situation.

What I would suggest is to rename the quota test `quota-tests-with-inflights`, and later we can have another test suite without, if needed... What do you think?

williamlardier replied:
(In ZKOP, the flag is enforced anyway)

francoisferrand replied:
> If we enable inflights by default, and we someday deploy a real scuba in the CI, we will still have debt when inflights are dropped.

In that case, we will actually be making a new change: which must be tested accordingly. In particular, having to disable in-flights when supporting a real scuba is a way to force adding these tests, as otherwise it would not work.

The debt/risk you are avoiding is not about cloudserver itself, but really about having it _silently_ enabled in cloudserver, and thus implicitely enabled... which would indeed break silently if that default value is changed later.

Since the flag is set anyway in ZKOP, we may keep the default in cloudserver to match the test: and be future-proof as Zenko/ZKOP will gracefully handle changes to this default (not very satifying as this is technically a breaking change, but risk should be managed already in this specific case...)

> If we disable inflights by default, this is the same situation.

If we disable inflights by default, but run tests only for another default: we introduce a false sense that the feature is present and tested... And when people start using it, there is a risk that no extra test is added: they will simply build on it, and if the code is good enough not notice the lack of (functional) test...

The default case is the default, so it must be most thoroughly tested. Hence either adding tests for that default case or switching the default.

williamlardier replied:
The mock was updated to support the no-inflight mode (S3 API calls will, in this case, directly update the mock, to simulate a system that would work without passing any inflights). At the end, the same tests are running for both scenarios.

francoisferrand commented on lib/api/apiUtils/quotas/quotaUtils.js:
```diff
@@ -0,0 +1,248 @@
+                // eslint-disable-next-line no-param-reassign
+                request.removeQuotaEvaluationInflights = {
+                    ...request.removeQuotaEvaluationInflights,
+                    [apiName]: removeInflights,
```
* it seems to be it may be simpler and cleaner to just introduce a callback called in any case at the end of the request : and let this callback decide if it needs to clear the inflights or not.
* what is the point of indexing by `apiName` ? we should have only one API name for each request...

williamlardier replied:
This was changed to a table as discussed: it also has the benefit of handling in a clean way multiple actions having the same name (e.g., during multi object deletes).

francoisferrand commented on lib/api/apiUtils/quotas/quotaUtils.js:
```diff
@@ -0,0 +1,248 @@
+                // eslint-disable-next-line no-param-reassign
+                request.removeQuotaEvaluationInflights = {
+                    ...request.removeQuotaEvaluationInflights,
+                    [apiName]: removeInflights,
```
can this really happen?
this is not a generic finalizer/hooking facility, but specific to quota: and no other place where we set it AFAICS...

williamlardier replied:
Closing as it was removed in favor of the table

francoisferrand commented on lib/quotas/scuba/wrapper.js:
```diff
@@ -0,0 +1,69 @@
+            if (data?.date) {
+                const date = new Date(data.date);
+                if (Date.now() - date.getTime() > this.maxStaleness) {
+                    throw new Error('Data is stale');
```
nit: `Data us stale, disabling quotas`

francoisferrand commented on lib/api/objectPutCopyPart.js:
```diff
@@ -45,6 +46,7 @@ function objectPutCopyPart(authInfo, request, sourceBucket,
         versionId: reqVersionId,
         getDeleteMarker: true,
         requestType: 'objectGet',
+        enableQuota: true,
```
should probably add a comment here, to explain why the quota check is forcefully enabled on `Get` and disabled on `Put`, contrary to expectations...

francoisferrand replied:
(same for `ObjectCopy`)

williamlardier replied:
Done in both files

francoisferrand commented on lib/api/apiUtils/object/objectRestore.js:
```diff
@@ -142,6 +143,12 @@ function objectRestore(metadata, mdUtils, userInfo, request, log, callback) {
+                // not affect the utilization metrics.
+                if (request.removeQuotaEvaluationInflights?.[mdValueParams]) {
+                    return request.removeQuotaEvaluationInflights?.[mdValueParams.requestType]?.(
+                        () => callback(null, 200, responseHeaders));
```
instead of calling the cleanup callback, it would be cleaner to forcefully disable the quota check from `standardMetadataValidateBucketAndObj` (for this case only, e.g. add `disableQuota: true` at line 63), and thus explicitely call `validateQuota()` from here only when needed...

williamlardier replied:
The API was updated: quota check is disabled during authz, then we evaluate it manually just before updating the MD, and if the object state matches. THis also ensures we return `InvalidObjectState` before `QuotaExceeded`. I had to rework the tests, so that we edit the object MD using a metadata client (here, mongodb) to really test a restore.

francoisferrand commented on lib/metadata/metadataUtils.js:
```diff
@@ -238,6 +238,16 @@ function standardMetadataValidateBucketAndObj(params, actionImplicitDenies, log,
         },
+        (bucket, objMD, next) => {
+            if (params.disableQuota || (!params.enableQuota && !requestType.some(type => actionNeedQuotaCheck[type] ||
+                actionWithDataDeletion[type]))) {
```
nit: API may be simpler with a single parameter (`checkQuota` ?) instead of two parameters (`disableQuota` and `enableQuota`)
* `checkQuota === undefined` : automatically depending on the action
* `checkQuota === true` : always check quota
* `checkQuota === false` : never check quota

```suggestion
            const needQuotaCheck = requestType -> requestType.some(type => actionNeedQuotaCheck[type] || actionWithDataDeletion[type]);
            const checkQuota = params.checkQuota === undefined ? needQuotaCheck(requestType) : params.checkQuota;
            if (!checkQuota) {
```

williamlardier replied:
Thanks, I cannot use your suggestion directly as I need to update the logic, but this works well indeed - PR was updated

francoisferrand commented on tests/unit/quotas/quotas.js:
empty file...

francoisferrand commented on lib/api/apiUtils/quotas/quotaUtils.js:
```diff
@@ -0,0 +1,231 @@
+    } else if (!bytes && objMD?.['content-length']) {
+        if (apiMethod === 'objectCopy') {
+            // object is being copied
+            bytes = Number.parseInt(objMD['content-length'], 10);
```
the target of the copy may already exist, and should be accounted for in non-versioned buckets.

and we actually have this case in the product, when updating metadata (from the UI) in case bucket is not versioned...

williamlardier replied:
You are right, I updated the logic by passing an optional parameter: `destObjMD`, known at the quota evaluation time for Copy APIs, so that we can, in the `processBytesToWrite` function, know if the copy is a replacement. This is unit tested.
Also added a filter on the copy part because it was missing

williamlardier replied:
It's in https://github.com/scality/cloudserver/pull/5554/commits/3ca848057cff1ddf55f357c24cb8759aad1c2f94

francoisferrand commented on lib/api/objectPutCopyPart.js:
```diff
@@ -45,6 +46,14 @@ function objectPutCopyPart(authInfo, request, sourceBucket,
+         * action. But in this context, the source object metadata is still
+         * unknown. In the context of quotas, to know the number of bytes that
+         * are being written, we explicitly enable the quota evaluation logic
+         * during the objectGet action.
```
nit:
```suggestion
         * during the objectGet action instead.
```

francoisferrand commented on lib/api/objectPutCopyPart.js:
```diff
@@ -88,6 +98,7 @@ function objectPutCopyPart(authInfo, request, sourceBucket,
         objectKey: destObjectKey,
         partNumber: paddedPartNumber,
         uploadId,
+        enableQuota: true,
```
why do we need a second quota check in the `uploadPartCopy` ?

williamlardier replied:
Upload part copy acts as an upload part, but the part is an existing object: we need to evaluate the quotas in the same way as copyObject during the `get` part, because we write a new object...
Or maybe I misunderstood your question?

williamlardier replied:
I added a missing filter on the copy part API

benzekrimaha commented on tests/unit/Config.js:
```diff
@@ -350,6 +350,92 @@ describe('Config', () => {
+            );
+        });
+    });
+
```
could be good to have a test where staleness is not an integer to make sure the error is thrown

williamlardier replied:
added a test that confirms that in this case, we fallback to the default value

benzekrimaha commented on tests/quota/awsNodeSdk.js:
```diff
@@ -252,7 +257,7 @@ function multiObjectDelete(bucket, keys, callback) {
 }
 
-describe('quota evaluation with scuba metrics', function t() {
+runIfMongo('quota evaluation with scuba metrics', function t() {
```
as runIfMongo not used anywhere else : 
`(process.env.S3METADATA === 'mongodb' ? describe : describe.skip)`

benzekrimaha commented on tests/unit/api/apiUtils/quotas/quotaUtils.js:
```diff
@@ -500,12 +500,40 @@ describe('processBytesToWrite', () => {
         assert.strictEqual(bytes, 100);
     });
+
+    it('should return content length of the bject metadata is the action is a copy part (getObject authz)', () => {
```
`should return content length of the bject metadata **if** the action is a copy part (getObject authz)`

benzekrimaha reviewed (approved):
LGTM (very minor none blocking comments as others have been adressed and the logic tested on the plateform as well), fixups commits should be squashed as well before merging

Seragonia committed:
CLDSRV-515: functional testing

Seragonia committed:
CLDSRV-515: unit testing

Seragonia committed:
CLDSRV-515: adapt dockerfile for scubaclient

ghost commented:
/approve

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/8.8`



The following branches have **NOT** changed:

* `development/7.10`
* `development/7.4`
* `development/7.70`
* `development/8.6`
* `development/8.7`



Please check the status of the associated issue CLDSRV-515.

Goodbye williamlardier.




*The following options are set:* **approve**
